File: rails/paybear.rb
require 'open-uri'
$PAYBEAR_SECRET = "secAPIKEY"

def get_address(order_id, token = 'ETH')
  token = token.downcase
  callback_url = "http://CHANGEME.com/callback/#{order_id}"
  url = "https://api.paybear.io/v2/#{token}/payment/#{CGI.escape(callback_url)}?token=#{$PAYBEAR_SECRET}"


  begin
    response = ActiveSupport::JSON.decode(open(url).read)
  rescue *HTTP_ERRORS => e
    {
        error: e,
        address: nil
    }
  else
    if response['success']
      {
          error: nil,
          address: response['data']['address']
      }
    end
  end
end

def get_currencies
  $currencies ||= nil

  if $currencies.nil?
    url = "https://api.paybear.io/v2/currencies?token=#{$PAYBEAR_SECRET}"
    response = ActiveSupport::JSON.decode(open(url).read)

    if response['success']
      $currencies = response['data']
    end
  end
  $currencies
end

def get_currency(token, order_id, get_address = false)
  rate = get_rate(token)

  if rate
    fiat_value = 19.99
    coins_value = (fiat_value / rate).round(8)

    token = token.downcase

    currencies = get_currencies
    if currencies[token]
      currency              = currencies[token]
      currency[:coinsValue] = coins_value

      if get_address
        currency[:address] = get_address(order_id, token)
      else
        currency[:currencyUrl] = "currencies/#{order_id}/#{token}"
      end
      currency
    end
  else
    puts "can't get rate for #{token}"
    nil
  end
end

def get_rate(cur_code)
  rates = get_rates
  cur_code = cur_code.downcase

  rates[cur_code] ? rates[cur_code]['mid'] : nil
end

def get_rates
  $rates ||= nil

  if $rates.nil?
    url = 'https://api.paybear.io/v2/exchange/usd/rate'
    response = ActiveSupport::JSON.decode(open(url).read)

    if response['success']
      $rates = response['data']
    end
  end
  $rates
end

File: php/paybear.php
<?php
define('PAYBEAR_SECRET', 'secAPIKEY'); //get your key at www.paybear.io

function getAddress($orderId, $token = 'ETH') {
	$callbackUrl = 'http://CHANGEME.com/callback.php?id='.$orderId;

	$token = strtolower($token);
	$url = sprintf('https://api.paybear.io/v2/%s/payment/%s?token=%s', $token, urlencode($callbackUrl), PAYBEAR_SECRET);
	if ($response = file_get_contents($url)) {
		$response = json_decode($response);
		if (isset($response->data->address)) {
			return $response->data->address;
		}
	}

	return null;
}

function getCurrencies()
{
	static $currencies = false;

	if (!$currencies) {
		$url = sprintf( "https://api.paybear.io/v2/currencies?token=%s", PAYBEAR_SECRET );
		if ( $response = @file_get_contents( $url ) ) {
			$response = json_decode( $response, true );
			if ( isset( $response ) && $response['success'] ) {
				$currencies = $response['data'];
			}
		}
	}

	return $currencies;
}

function getCurrency($token, $orderId, $getAddress = false) {
	$rate = getRate($token);

	if ($rate) {
		$fiatValue = 19.99; //get from $orderId
		$coinsValue = round($fiatValue / $rate, 8);

		$token = strtolower($token);

		$currencies = getCurrencies();
		if (isset($currencies[$token])) {
			$currency               = $currencies[ $token ];
			$currency['coinsValue'] = $coinsValue;
			$currency['rate'] = $rate;

			if ( $getAddress ) {
				$currency['address'] = getAddress( $orderId, $token );
			} else {
				$currency['currencyUrl'] = sprintf( 'currencies.php?order=%s&token=%s', $orderId, $token );
			}

			return $currency;
		}

	}

	echo 'can\'t get rate for '.$token;
	return null;
}


function getRate($curCode) {
	$rates = getRates();
	$curCode = strtolower($curCode);

	return isset($rates->$curCode) ? $rates->$curCode->mid : false;
}

function getRates() {
	static $rates = null;

	if (empty($rates)) {
		$url = "https://api.paybear.io/v2/exchange/usd/rate";

		if ( $response = file_get_contents( $url ) ) {
			$response = json_decode( $response );
			if ( $response->success ) {
				$rates = $response->data;
			}
		}
	}
	return $rates;
}


/*
function getRate($curCode) {
    $url = "https://api.paybear.io/v2/".strtolower($curCode)."/exchange/usd/rate";

    if ($response = file_get_contents($url)) {
        $response = json_decode($response);
        if ($response->success) {
            return $response->data->mid;
        }
    }
    return null;
}

*/